### PR TITLE
[oneDNN v3.x]: Added support for batchnorm fwd and bwd for fp32 and bf16

### PR DESCRIPTION
**Pull request by bhavani-subramanian**
- This PR adds support for oneDNN v3.x in batchnorm fwd and bwd for fp32 and bf16.
- This PR passes all unit tests related to batchnorm fwd and bwd for fp32/bf16 when oneDNN v3.x is enabled.
- This PR does not add or affect Eigen ops and is specific only to oneDNN kernels.
- For additional comments related to oneDNN v3.x integration, please refer to [this](https://github.com/tensorflow/tensorflow/pull/60307#issue-1665423762) comment.

### CODE REVIEW

**Comment by gbaned** [comment]
Hi @penpornk Can you please review this PR ? Thank you!

**Comment by gbaned** [comment]
Hi @penpornk Can you please review this PR ? Thank you!

File: tensorflow/core/kernels/mkl/mkl_fused_batch_norm_op.cc [hunk 1]
```diff
@@ -548,18 +674,25 @@ class MklFusedBatchNormBwdPrimitive : public MklPrimitive {
...
     // Forward batch-normalization descriptor and primitive descriptor.
     // Adding this back due to type difference with context.flags
-    auto bn_flags = bwdParams.training
-                        ? dnnl::normalization_flags::use_scale_shift
-                        : (dnnl::normalization_flags::use_scale_shift |
-                           dnnl::normalization_flags::use_global_stats);
+    auto bn_flags =
+        bwdParams.training
+            ? GET_SCALE_AND_SHIFT_FLAGS
+            : (GET_SCALE_AND_SHIFT_FLAGS | GET_FLAG(use_global_stats));
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl/mkl_fused_batch_norm_op.cc:691]: This is the 3rd time I see this logic. Can we refactor this to an inline function, e.g., `bn_flags = GetBatchNormFlag(bwdParams.training, USE_SCALE_AND_SHIFT, use_global_stats);` ?

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for the PR and sorry for the delay!

File: tensorflow/core/kernels/mkl/mkl_fused_batch_norm_op.cc [hunk 2]
```diff
@@ -77,27 +113,32 @@ class MklFusedBatchNormFwdPrimitive : public MklPrimitive {
...
 
   // BatchNormalization forward execute
-  //   src_data:     input data buffer of src
-  //   weights_data: input data buffer of weights
-  //   dst_data:     output data buffer of dst
-  //   mean_data:     output data buffer of means
-  //   variance_data: output data buffer of variances
-  void Execute(const T* src_data, const U* weights_data, T* dst_data,
+  //   src_data:         input data buffer of src
+  //   scale_shift_data: input data buffer of scale_shift
```
[the remark this answers is not in the document]
Comment (bhavani-subramanian) [reply tensorflow/core/kernels/mkl/mkl_fused_batch_norm_op.cc]: Updated.

File: tensorflow/core/kernels/mkl/mkl_fused_batch_norm_op.cc [hunk 3]
```diff
@@ -195,35 +260,52 @@ class MklFusedBatchNormFwdPrimitive : public MklPrimitive {
   void Setup(const MklBatchNormFwdParams& fwdParams) {
     context_.flags =
         fwdParams.training
-            ? GET_FLAG(use_scale_shift)
-            : (GET_FLAG(use_scale_shift) | GET_FLAG(use_global_stats));
-    context_.pkind = fwdParams.training ? prop_kind::forward_training
-                                        : prop_kind::forward_scoring;
+            ? GET_SCALE_AND_SHIFT_FLAGS
+            : (GET_SCALE_AND_SHIFT_FLAGS | GET_FLAG(use_global_stats));
```
[the remark this answers is not in the document]
Comment (bhavani-subramanian) [reply tensorflow/core/kernels/mkl/mkl_fused_batch_norm_op.cc]: Makes sense. Updated.

File: tensorflow/core/kernels/mkl/mkl_fused_batch_norm_op.cc [hunk 4]
```diff
@@ -530,16 +650,22 @@ class MklFusedBatchNormBwdPrimitive : public MklPrimitive {
...
           diff_src_mem(nullptr) {}
   };
 
   void Setup(const MklBatchNormBwdParams& bwdParams) {
     context_.flags =
         bwdParams.training
-            ? GET_FLAG(use_scale_shift)
-            : (GET_FLAG(use_scale_shift) | GET_FLAG(use_global_stats));
+            ? GET_SCALE_AND_SHIFT_FLAGS
+            : (GET_SCALE_AND_SHIFT_FLAGS | GET_FLAG(use_global_stats));
```
[the remark this answers is not in the document]
Comment (bhavani-subramanian) [reply tensorflow/core/kernels/mkl/mkl_fused_batch_norm_op.cc]: Done.

File: tensorflow/core/kernels/mkl/mkl_fused_batch_norm_op.cc (hunk 1 above)
Comment (bhavani-subramanian) [reply tensorflow/core/kernels/mkl/mkl_fused_batch_norm_op.cc]: Yes, fixed it.

**Comment by bhavani-subramanian** [comment]
@penpornk I have addressed your review comments. Please take a look. Thanks.

File: tensorflow/core/kernels/mkl/mkl_concat_op.cc [hunk 5]
```diff
@@ -364,7 +364,7 @@ class MklConcatFwdPrimitive : public MklPrimitive {
              const std::vector<memory::desc>& srcs_md) {
     // Create memory descriptors for concat with specified srcs format
     for (size_t i = 0; i < concat_fwd_dims.num_inputs; i++) {
-      dnnl::memory::desc source_md(memory::desc(GET_MEMORY_DESC(srcs_md[i])));
+      dnnl::memory::desc source_md((memory::desc(GET_MEMORY_DESC(srcs_md[i]))));
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl/mkl_concat_op.cc:367]: What clang error did we get without the extra pair of parentheses? Could you please post the error message? Thank you!

Comment (bhavani-subramanian) [reply tensorflow/core/kernels/mkl/mkl_concat_op.cc:367]: @penpornk Below is the error from Clang:
```
tensorflow/core/kernels/mkl/mkl_concat_op.cc:367:35: warning: parentheses were disambiguated as a function declaration [-Wvexing-parse]
      dnnl::memory::desc source_md(memory::desc(GET_MEMORY_DESC(srcs_md[i])));
                                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
tensorflow/core/kernels/mkl/mkl_concat_op.cc:367:36: note: add a pair of parentheses to declare a variable
      dnnl::memory::desc source_md(memory::desc(GET_MEMORY_DESC(srcs_md[i])));
                                   ^
                                   (                                        )
tensorflow/core/kernels/mkl/mkl_concat_op.cc:368:23: error: no matching member function for call to 'push_back'
      context_.src_md.push_back(source_md);
      ~~~~~~~~~~~~~~~~^~~~~~~~~
```

Comment (penpornk) [reply tensorflow/core/kernels/mkl/mkl_concat_op.cc:367]: Got it. Thank you very much!

**Review by penpornk — APPROVED** [review]
